src/search/algorithms/named_vector.h

```diff
@@ -0,0 +1,103 @@
+#ifndef ALGORITHMS_NAMED_VECTOR_H
+#define ALGORITHMS_NAMED_VECTOR_H
+
+#include <string>
+#include <vector>
+
+namespace named_vector {
+/*
+  NamedVector is a vector-like collection with optional names to be associated with each element.
+  It is optimized to have minimal overhead when there are no names.
+  Any name which is not specified is assumed to be an empty string.
+ */
+template<typename T>
```

Comment:
Please document that the intended use of `NamedVector` is for debugging, i.e., where we wanted names for debugging that can be set only in debugging mode without causing an overhead when they are not set. (This could expand or replace the second sentence in documentation).

Also, please document that accessing the name of an invalid index is an error (see below).